File: memory_hub_app/lib/models/family/family_calendar.dart
class FamilyCalendarEvent {
  final String id;
  final String title;
  final String? description;
  final DateTime startDate;
  final DateTime? endDate;
  final bool isAllDay;
  final String? location;
  final String eventType;
  final String? recurrenceRule;
  final List<String> familyCircleIds;
  final List<String> attendeeIds;
  final String? reminder;
  final String createdBy;
  final String? createdByName;
  final DateTime createdAt;
  final DateTime updatedAt;

  FamilyCalendarEvent({
    required this.id,
    required this.title,
    this.description,
    required this.startDate,
    this.endDate,
    required this.isAllDay,
    this.location,
    required this.eventType,
    this.recurrenceRule,
    required this.familyCircleIds,
    required this.attendeeIds,
    this.reminder,
    required this.createdBy,
    this.createdByName,
    required this.createdAt,
    required this.updatedAt,
  });

  factory FamilyCalendarEvent.fromJson(Map<String, dynamic> json) {
    return FamilyCalendarEvent(
      id: json['id'] ?? json['_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'],
      startDate: DateTime.parse(json['start_date'] ?? DateTime.now().toIso8601String()),
      endDate: json['end_date'] != null
          ? DateTime.parse(json['end_date'])
          : null,
      isAllDay: json['is_all_day'] ?? false,
      location: json['location'],
      eventType: json['event_type'] ?? 'general',
      recurrenceRule: json['recurrence_rule'],
      familyCircleIds: List<String>.from(json['family_circle_ids'] ?? []),
      attendeeIds: List<String>.from(json['attendee_ids'] ?? []),
      reminder: json['reminder'],
      createdBy: json['created_by'] ?? '',
      createdByName: json['created_by_name'],
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
    );
  }
}

File: app/models/family/family_calendar.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
from bson import ObjectId
from enum import Enum

from app.models.user import PyObjectId


class EventType(str, Enum):
    BIRTHDAY = "birthday"
    ANNIVERSARY = "anniversary"
    DEATH_ANNIVERSARY = "death_anniversary"
    GATHERING = "gathering"
    HOLIDAY = "holiday"
    REMINDER = "reminder"
    HISTORICAL_EVENT = "historical_event"
    OTHER = "other"


class EventRecurrence(str, Enum):
    NONE = "none"
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    YEARLY = "yearly"


class FamilyEventBase(BaseModel):
    title: str
    description: Optional[str] = None
    event_type: EventType
    event_date: datetime
    end_date: Optional[datetime] = None
    location: Optional[str] = None
    recurrence: EventRecurrence = EventRecurrence.NONE


class FamilyEventCreate(FamilyEventBase):
    family_circle_ids: List[str] = Field(default_factory=list)
    attendee_ids: List[str] = Field(default_factory=list)
    reminder_minutes: Optional[int] = None
    genealogy_person_id: Optional[str] = None
    auto_generated: bool = False


class FamilyEventUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    event_type: Optional[EventType] = None
    event_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    location: Optional[str] = None
    recurrence: Optional[EventRecurrence] = None
    family_circle_ids: Optional[List[str]] = None
    attendee_ids: Optional[List[str]] = None
    reminder_minutes: Optional[int] = None


class FamilyEventInDB(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    title: str
    description: Optional[str] = None
    event_type: EventType
    event_date: datetime
    end_date: Optional[datetime] = None
    location: Optional[str] = None
    recurrence: EventRecurrence
    created_by: PyObjectId
    family_circle_ids: List[PyObjectId] = Field(default_factory=list)
    attendee_ids: List[PyObjectId] = Field(default_factory=list)
    reminder_minutes: Optional[int] = None
    reminder_sent: bool = False
    genealogy_person_id: Optional[PyObjectId] = None
    auto_generated: bool = False
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Config:
        json_encoders = {ObjectId: str}
        populate_by_name = True
        arbitrary_types_allowed = True


class FamilyEventResponse(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    event_type: EventType
    event_date: datetime
    end_date: Optional[datetime] = None
    location: Optional[str] = None
    recurrence: EventRecurrence
    created_by: str
    created_by_name: Optional[str] = None
    family_circle_ids: List[str]
    attendee_ids: List[str]
    attendee_names: List[str] = Field(default_factory=list)
    reminder_minutes: Optional[int] = None
    genealogy_person_id: Optional[str] = None
    genealogy_person_name: Optional[str] = None
    auto_generated: bool = False
    created_at: datetime
    updated_at: datetime
